### PR TITLE
Format `assert` statement

### PR DESCRIPTION
**Pull request by cnpryer**
## Summary

Format `assert` statement with `Parenthesize::IfBreaks` for test and message expressions.

## Test Plan

Add ruff fixture with comments.

There's some coverage already as well including:

- `assert left > right`
- `assert test, msg`

I'll try to leave the *good first issue* ones alone now :)

This looks really good btw.

-----

TODO (required):
- [x] Ruff fixture
- [x] Formatting [`test`'s value](https://github.com/astral-sh/ruff/pull/5168#issuecomment-1631148612)
- [ ] Formatting [`msg` long strings](https://github.com/astral-sh/ruff/pull/5168#issuecomment-1630767421) UPDATE: [comment](https://github.com/astral-sh/ruff/pull/5168#discussion_r1262709467)
- [x] Updated snapshots

TODO (extra):
- [ ] Handling [misplaced trailing own-line comments](https://github.com/astral-sh/ruff/pull/5168#issuecomment-1630767421) UPDATE: [comment](https://github.com/astral-sh/ruff/pull/5168#discussion_r1262740871)

### CODE REVIEW

**Comment by cnpryer** [comment]
Hmm maybe there's more here. This isn't formatting the same as black.

```py
assert (
    # Some comment
    True  # Some comment
    # Some comment
), "Some string"  # Some comment
# Some comment
```

File: crates/ruff_python_formatter/resources/test/fixtures/black/simple_cases/assert.py
Comment (konstin) [inline crates/ruff_python_formatter/resources/test/fixtures/black/simple_cases/assert.py:1]: these test cases are copied from black, please add new test cases to `crates/ruff_python_formatter/resources/test/fixtures/ruff` instead

Comment (cnpryer) [reply crates/ruff_python_formatter/resources/test/fixtures/black/simple_cases/assert.py:1]: d780f2f8fa6dba9234e9b9301e1eae2fddddf5cd

**Comment by cnpryer** [comment]
> Hmm maybe there's more here. This isn't formatting the same as black.
> 
> ```python
> assert (
>     # Some comment
>     True  # Some comment
>     # Some comment
> ), "Some string"  # Some comment
> # Some comment
> ```

Reported this [here](https://github.com/psf/black/issues/638#issuecomment-1597665572) for now. Probably should be a new issue if that one it's mentioned in is not updated. I'll get back to this.

**Comment by cnpryer** [comment]
I find this snapshot result odd, but it could just be my inexperience with `insta`.
```
Reviewing [1/11] ruff_python_formatter@0.0.0:
Snapshot file: crates/ruff_python_formatter/tests/snapshots/black_compatibility@simple_cases__expression.py.snap
Snapshot: expression
Source: crates/ruff_python_formatter/tests/fixtures.rs:86
Input file: crates/ruff_python_formatter/resources/test/fixtures/black/simple_cases/expression.py
────────────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────────────
  391   391 │ -    str,
  392   392 │ -    int,
  393   393 │ -    float,
  394   394 │ -    dict[str, int],
        395 │+-]
        396 │+-very_long_variable_name_filters: t.List[
        397 │+-    t.Tuple[str, t.Union[str, t.List[t.Optional[str]]]],
  395   398 │ +    (
  396   399 │ +        str,
  397   400 │ +        int,
  398   401 │ +        float,
  399   402 │ +        dict[str, int],
  400   403 │ +    )
  401   404 │  ]
  402       │--very_long_variable_name_filters: t.List[
  403       │--    t.Tuple[str, t.Union[str, t.List[t.Optional[str]]]],
  404       │--]
  405   405 │ -xxxx_xxxxx_xxxx_xxx: Callable[..., List[SomeClass]] = classmethod(  # type: ignore
  406   406 │ -    sync(async_xxxx_xxx_xxxx_xxxxx_xxxx_xxx.__func__)
  407   407 │ -)
  408   408 │ -xxxx_xxx_xxxx_xxxxx_xxxx_xxx: Callable[..., List[SomeClass]] = classmethod(  # type: ignore
```

**Comment by MichaReiser** [comment]
> I find this snapshot result odd, but it could just be my inexperience with `insta`.
> 
> ```
> Reviewing [1/11] ruff_python_formatter@0.0.0:
> Snapshot file: crates/ruff_python_formatter/tests/snapshots/black_compatibility@simple_cases__expression.py.snap
> Snapshot: expression
> Source: crates/ruff_python_formatter/tests/fixtures.rs:86
> Input file: crates/ruff_python_formatter/resources/test/fixtures/black/simple_cases/expression.py
> ────────────────────────────────────────────────────────────────────────────────────────
> -old snapshot
> +new results
> ────────────┬───────────────────────────────────────────────────────────────────────────
>   391   391 │ -    str,
>   392   392 │ -    int,
>   393   393 │ -    float,
>   394   394 │ -    dict[str, int],
>         395 │+-]
>         396 │+-very_long_variable_name_filters: t.List[
>         397 │+-    t.Tuple[str, t.Union[str, t.List[t.Optional[str]]]],
>   395   398 │ +    (
>   396   399 │ +        str,
>   397   400 │ +        int,
>   398   401 │ +        float,
>   399   402 │ +        dict[str, int],
>   400   403 │ +    )
>   401   404 │  ]
>   402       │--very_long_variable_name_filters: t.List[
>   403       │--    t.Tuple[str, t.Union[str, t.List[t.Optional[str]]]],
>   404       │--]
>   405   405 │ -xxxx_xxxxx_xxxx_xxx: Callable[..., List[SomeClass]] = classmethod(  # type: ignore
>   406   406 │ -    sync(async_xxxx_xxx_xxxx_xxxxx_xxxx_xxx.__func__)
>   407   407 │ -)
>   408   408 │ -xxxx_xxx_xxxx_xxxxx_xxxx_xxx: Callable[..., List[SomeClass]] = classmethod(  # type: ignore
> ```

The snapshots can be hard to read because it shows a diff of the new to old snapshot containing a diff with the black/ruff formatter differences. 

Overall: 
* Red lines / Lines gone -> Good (not very intuitive)
* More lines with a single `+` or `-` -> Good. It means the output of black and ruff are the same 
* New lines with `++` or `+-`: Potentially a regression, but may be fine. Check out the section below that show the black and ruff output on their own and compare it manually.

**Comment by cnpryer** [comment]
Thanks for the explanation. Yea I that's tough to understand. Manually reviewing it that result looks fine. Intuitively I would have expected no change in the diff here, but I'm sure I'm just missing something.

**Comment by cnpryer** [comment]
This is maybe a better example of what's tripping me up. Manually reviewing the new snapshot ruff and black outputs, and comparing them with the original snapshot, I'd expect there to be no new results displayed during the `insta` review because everything seems to be fine. Instead I get:
 

```
black_compatibility@simple_cases__fmtonoff2.py.snap
   71    71 │      pass
   72    72 │  
   73    73 │ +
   74    74 │  def check_fader(test):
   75       │-+    pass
         75 │+-
         76 │+     pass
   76    77 │  
   77       │--    pass
   78       │- 
         78 │++
```

IIUC this result is displaying the same diff just in a different way.

Originally
```
+
 def check_fader(test):
+    pass
 
-    pass
```

And in this result
```
+
 def check_fader(test):
-
     pass
```

**Comment by MichaReiser** [comment]
> This is maybe a better example of what's tripping me up. Manually reviewing the new snapshot ruff and black outputs, and comparing them with the original snapshot, I'd expect there to be no new results displayed during the `insta` review because everything seems to be fine. Instead I get:
> 
> ```
> black_compatibility@simple_cases__fmtonoff2.py.snap
>    71    71 │      pass
>    72    72 │  
>    73    73 │ +
>    74    74 │  def check_fader(test):
>    75       │-+    pass
>          75 │+-
>          76 │+     pass
>    76    77 │  
>    77       │--    pass
>    78       │- 
>          78 │++
> ```
> 
> IIUC this result is displaying the same diff just in a different way.
> 
> Originally
> 
> ```
> +
>  def check_fader(test):
> +    pass
>  
> -    pass
> ```
> 
> And in this result
> 
> ```
> +
>  def check_fader(test):
> -
>      pass
> ```

Yeah this is a bit annoying and something that I noticed too. The diff algorithm sometimes gets dripped up and produces changes even if there are non (to this particular section). We could explore if other diffing algorithm help but ultimately the best is to have higher compatibility, which will make the diff algorithm's job easier.

**Review by MichaReiser — APPROVED** [review]
This looks ready to merge for me. Is there something that you want to look into before merging?

File: crates/ruff_python_formatter/src/statement/stmt_assert.rs
Comment (cnpryer) [reply crates/ruff_python_formatter/src/statement/stmt_assert.rs:38]: For sure!

Comment (cnpryer) [reply crates/ruff_python_formatter/src/statement/stmt_assert.rs]: aaffdac1224d3217bc40371832f0a174e0d2a447

Comment (cnpryer) [reply crates/ruff_python_formatter/src/statement/stmt_assert.rs]: Tbh I don't even see this as a nit lol.

**Comment by cnpryer** [comment]
Hmm there's more to this than what I mentioned. I added `optional_parentheses` for the formatted `msg`, but what I really need to do is consider the entire statement, formatting `test`'s value first prior to having the `msg` break with parentheses.

```
        488 │+-assert this is ComplexTest and not requirements.fit_in_a_single_line(
        489 │+-    force=False
        490 │+-), "Short message"
        491 │++assert this is ComplexTest and not requirements.fit_in_a_single_line(force=False), (
        492 │++    "Short message"
        493 │++)
        494 │+ assert parens is TooMany
        495 │+ for (x,) in (1,), (2,), (3,):
```

`black` would format `fit_in_a_single_line`, making it unnecessary to futher format the `msg`.

File: crates/ruff_python_formatter/src/statement/stmt_assert.rs
Comment (cnpryer) [reply crates/ruff_python_formatter/src/statement/stmt_assert.rs]: I ended up adding a `group`, so I've deviated from your suggestion. I'm still finding my feet with some of this stuff, so feel free to tell me to stop over-implementing this.

I'd want to format this next
```py
# Regression test for https://github.com/psf/black/issues/3414.
assert xxxxxxxxx.xxxxxxxxx.xxxxxxxxx(
    xxxxxxxxx
).xxxxxxxxxxxxxxxxxx(), (
    "xxx {xxxxxxxxx} xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
)
```
And then finally figure out comment placement, but I can rip out the fixture I've added for that and follow up after this is merged.

**Comment by MichaReiser** [comment]
I don't have a good answer for you yet but am looking into a similar problem with `with` statements (the optional parentheses there are tricky). 

My understanding from [Black's code](https://github.com/psf/black/blob/d1248ca9beaf0ba526d265f4108836d89cf551b7/src/black/linegen.py#L484-L499) is that `assert` should work the same as for `if` headers. Now, if our formatting of the if header is correct is another question...

File: crates/ruff_python_formatter/src/statement/stmt_assert.rs
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/statement/stmt_assert.rs:29]: Is this different from `msg.format().with_options(Parenthesize::IfBreaks)`?

Comment (cnpryer) [reply crates/ruff_python_formatter/src/statement/stmt_assert.rs:29]: Yea I'm definitely not satisfied with it yet.

Without my "hack" here I would get 

```
black_compatibility@simple_cases__composition_no_trailing_comma.py.snap
  349       │-        } == expected, (
  350       │-            "Not what we expected and the message is too long to fit in one line"
  351       │-        )
        360 │+        } == expected, "Not what we expected and the message is too long to fit in one line"
```
`black` will wrap the `msg` (lines 349-351). So IIUC `Parenthesize::IfBreaks` wouldn't give me that alone since the `msg` doesn't break as a long string like that.

> It may also be that e.g. this is caused by incorrect formatting of string

I don't know enough yet but maybe this is possible.

Comment (cnpryer) [reply crates/ruff_python_formatter/src/statement/stmt_assert.rs:29]: > My understanding from [Black's code](https://github.com/psf/black/blob/d1248ca9beaf0ba526d265f4108836d89cf551b7/src/black/linegen.py#L484-L499) is that assert should work the same as for if headers

This could be helpful. I'll check it out. Thank you!

Comment (MichaReiser) [reply crates/ruff_python_formatter/src/statement/stmt_assert.rs:29]: Let's see if https://github.com/astral-sh/ruff/pull/5708 helps

Comment (MichaReiser) [reply crates/ruff_python_formatter/src/statement/stmt_assert.rs:29]: I spent some more time reading black's source code and what I think we miss is that strings have a higher split priority than any math operator. I'm not sure how we want to model this yet but I suggest to not block this PR because of it.

**Comment by cnpryer** [comment]
Did a skim of the updated snapshots. I can check more thoroughly after work, but for now I'll mark this as ready for a review.

File: crates/ruff_python_formatter/tests/snapshots/format@statement__assert.py.snap
Comment (cnpryer) [inline crates/ruff_python_formatter/tests/snapshots/format@statement__assert.py.snap:73]: I thought it'd be better to leave this here instead of ripping it out entirely.

**Comment by cnpryer** [comment]
Got it

File: crates/ruff_python_formatter/src/statement/stmt_assert.rs
Comment (cnpryer) [reply crates/ruff_python_formatter/src/statement/stmt_assert.rs]: 🎉 709c4a70f6d4ba6ce91d12a71cde94dee34789a9

Comment (MichaReiser) [reply crates/ruff_python_formatter/src/statement/stmt_assert.rs]: Nice. Yeah, it can sometimes be very subtle.

File: crates/ruff_python_formatter/tests/snapshots/black_compatibility@miscellaneous__long_strings_flag_disabled.py.snap
Comment (MichaReiser) [inline crates/ruff_python_formatter/tests/snapshots/black_compatibility@miscellaneous__long_strings_flag_disabled.py.snap:326]: Ehh, this is interesting. It seems black happily wraps identifiers in parentheses in *some* positions but not all. This is fine for now.

**Review by MichaReiser — APPROVED** [review]
Perfect! Thank you so much. This will increase our coverage significantly.

File: crates/ruff_python_formatter/tests/snapshots/black_compatibility@miscellaneous__long_strings_flag_disabled.py.snap
Comment (davidszotten) [reply crates/ruff_python_formatter/tests/snapshots/black_compatibility@miscellaneous__long_strings_flag_disabled.py.snap:326]: i wonder if that's because assert is somewhat unusual in that one must not wrap the entire rhs in parens, i.e. 

`assert foo, "msg"` very much != `assert (foo, "msg")`

**Pull request merged**
